substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+					if let Some(stake) = approvals_map.get_mut(&nomination) {
+						*stake += score as sp_npos_elections::ExtendedBalance;
+					} else {
+						approvals_map.insert(nomination, score.into());
+					}
+				}
+			} else {
+				// if it is in the voter list but it's not a nominator, it should be a validator
+				// and part of the target list.
+				frame_support::ensure!(
+					T::Staking::status(&voter) == Ok(StakerStatus::Validator),
+					"wrong state of voter"
+				);
+				frame_support::ensure!(
+					T::TargetList::contains(&voter),
+					"if voter is in voter list and it's not a nominator, it must be a target"
+				);
+			}
+		}
+
+		// add self-vote of active targets to calculated approvals from the `TargetList` POV.
+		for target in T::TargetList::iter() {
+			// also checks invariant: all active targets are also voters.
+			let maybe_self_stake = match T::Staking::status(&target) {
+				Err(_) => {
+					// if target is "dangling" (i.e unbonded but still in the `TargetList`), it
+					// should NOT be part of the voter list.
+					frame_support::ensure!(
+						!T::VoterList::contains(&target),
+						"dangling target (i.e. unbonded) should not be part of the voter list"
+					);
+
+					// if target is dangling, its target score should > 0 (otherwise it should
+					// have been removed from the list).
+					frame_support::ensure!(
+                            T::TargetList::get_score(&target).expect("target must have score") > Zero::zero(),
+                            "dangling target (i.e. unbonded) is part of the `TargetList` IFF it's approval voting > 0"
+                        );
+					// no self-stake and it should not be part of the target list.
+					None
+				},
+				Ok(StakerStatus::Idle) => {
+					// target is idle and not part of the voter list.
+					frame_support::ensure!(
+						!T::VoterList::contains(&target),
+						"chilled validator (idle target) should not be part of the voter list"
+					);
+
+					// no sef-stake but since it's chilling, it should be part of the TL even
+					// with score = 0.
+					Some(0)
+				},
+				Ok(StakerStatus::Validator) => {
+					// active target should be part of the voter list.
+					frame_support::ensure!(
+						T::VoterList::contains(&target),
+						"bonded and active validator should also be part of the voter list"
+					);
+					// return self-stake (ie. active bonded).
+					T::Staking::stake(&target).map(|s| Self::weight_of(s.active)).ok()
+				},
+				Ok(StakerStatus::Nominator(_)) => {
+					panic!("staker with nominator status should not be part of the target list");
+				},
+			};
+
+			if let Some(score) = maybe_self_stake {
+				if let Some(stake) = approvals_map.get_mut(&target) {
+					*stake += score as sp_npos_elections::ExtendedBalance;
+				} else {
+					approvals_map.insert(target, score.into());
+				}
+			} else {
+				// unbonded target: it does not have self-stake.
+			}
+		}
+
+		log!(trace, "try-state: calculated approvals map: {:?}", approvals_map);
```

Comment:
boy this is gonna be a fun print ot look at :))

Reply:
removed!